pkg/pillar/pubsub/driver.go

```diff
@@ -0,0 +1,58 @@
+package pubsub
+
+// Driver a backend driver for pubsub
+type Driver interface {
+	// Publisher return a `DriverPublisher` for the given name and topic.
+	// The caller passes the `Updaters`, `Restarted` checker and `Differ`.
+	// These can be used to:
+	// * add to or remove from the updaters
+	// * determine if the topic has been restarted
+	// * diff the current known state from the target known state
+	Publisher(global bool, name, topic string, persistent bool, updaterList *Updaters, restarted Restarted, differ Differ) (DriverPublisher, error)
+	// Subscriber return a `DriverSubscriber` for the given name and topic.
+	// This is expected to create a `DriverSubscriber`, but not start it.
+	// Once started, when changes arrive, they should be published to the provided
+	// channel. Each update to the channel is of type `Change`, which encapsulates
+	// the operation, key and value.
+	Subscriber(global bool, name, topic string, persistent bool, C chan Change) (DriverSubscriber, error)
+}
+
+// DriverSubscriber interface that a driver for subscribing must implement
+type DriverSubscriber interface {
+	// Start subscribing to a name and topic and publish changes to the channel
+	Start() error
+	// DefaultName Return the default name to use for an agent, when the name
+	// is not provided.
+	DefaultName() string
+}
+
+// DriverPublisher interface that a driver for publishing must implement
+type DriverPublisher interface {
+	// Start the publisher
+	Start() error
+	// Load current status from persistence. Usually called only on first start.
+	// The implementation is responsible for determining if the load is necessary
+	// or already has been performed. If it has been already, it should not change
+	// anything. The caller has no knowledge of where the persistent state was
+	// stored: disk, databases, or vellum. All it cares about is that it gets
+	// a key-value list.
+	Load() (map[string]interface{}, bool, error)
+	// Publish a key-value pair to all subscribers and optionally persistence
+	Publish(key string, item interface{}) error
+	// Unpublish a key, i.e. delete it and publish its deletion to all subscribers
+	Unpublish(key string) error
+	// Restart set the state of the topic to restarted, or cancel the restarted
+	// state
+	Restart(restarted bool) error
```

Comment:
Do  we need both start and restart?

Reply:
Yes. The original implementation has both a stage where it starts publishing, as well as a way to flag that a topic is restarted, as well as clear the restarted flag. I never really understood what it is for, but accepted the logic. Maybe @eriknordmark can explain.

Reply:
Yes. The original implementation has both a stage where it starts publishing, as well as a way to flag that a topic is restarted, as well as clear the restarted flag. I never really understood what it is for, but accepted the logic. Maybe @eriknordmark can explain.

Reply:
I think part of the "restarted" logic was to support agent restart in future. Right now, it is there to indicate that the publisher is ready..